packages/tasit-action/src/contract/Action.js

```diff
@@ -7,54 +7,59 @@ import ConfigLoader from "../ConfigLoader";
 export class Action extends Subscription {
   #provider;
   #signer;
-  #rawTx;
+  #rawAction;
   #tx;
   #txConfirmations;
   #timeout;
   #lastConfirmationTime;
 
-  constructor(rawTx, provider, signer) {
+  constructor(rawAction, provider, signer) {
     // Provider implements EventEmitter API and it's enough
     //  to handle with transactions events
     super(provider);
 
     const { events } = ConfigLoader.getConfig();
     const { timeout } = events;
 
-    this.#rawTx = rawTx;
+    this.#rawAction = rawAction;
     this.#signer = signer;
     this.#timeout = timeout;
     this.#provider = provider;
     this.#txConfirmations = 0;
   }
 
-  _toRaw = async () => {
-    return await this.#rawTx;
+  _toRaw = () => {
+    return this.#rawAction;
   };
 
   #signAndSend = async () => {
-    // TODO: Go deep on gas handling.
-    // Without that, VM returns a revert error instead of out of gas error.
-    // See: https://github.com/tasitlabs/TasitSDK/issues/173
-    //
-    // This command isn't enough
-    // const gasLimit = await this.#provider.estimateGas(this.#rawTx);
-    const gasParams = {
-      gasLimit: 7e6,
-      gasPrice: 1e9,
-    };
+    try {
+      // Note: Resolving promise if the Action was created using a async rawTx
+      let rawTx = await this.#rawAction;
 
-    const nonce = await this.#provider.getTransactionCount(
-      this.#signer.address
-    );
+      const nonce = await this.#provider.getTransactionCount(
+        this.#signer.address
+      );
+      const gasPrice = await this.#provider.getGasPrice();
+      const network = await this.#provider.getNetwork();
+      const chainId = network.chainId;
+      let { value } = rawTx;
+      value = !value ? 0 : value;
 
-    let rawTx = await this.#rawTx;
+      rawTx = { ...rawTx, nonce, gasPrice, chainId, value };
 
-    rawTx = { ...rawTx, nonce, ...gasParams };
+      // TODO: Go deep on gas handling.
+      // Without that, VM returns a revert error instead of out of gas error.
+      // See: https://github.com/tasitlabs/TasitSDK/issues/173
+      //
+      // This command isn't enough
+      //const gasLimit = await this.#provider.estimateGas(rawTx);
+      const gasLimit = 7e6;
 
-    const signedTx = await this.#signer.sign(rawTx);
+      this.#rawAction = { ...rawTx, gasLimit };
+
+      const signedTx = await this.#signer.sign(this.#rawAction);
```

Comment:
Extract some of this logic out into a sign function that is then used in `signAndSend`?

Reply:
Working on that right now.

Reply:
I've created the `#fillRawAction` function.

A question: Is `fulfillRawAction` a better name?

Reply:
Hmm, what do you intend for `fill` or `fulfill` to mean? Is it just signing? If so, maybe `signRawAction` would be better. Or what else is happening that you want to convey in the name?

Reply:
Sorry, I haven't be cleared. Since sign is only one line of this function, I've thought that is better to extract the logic related to fill the rawTransaction (with gas params, nonce, value, etc).
How is the function is now:

```javascript
#signAndSend = async () => {
  try {
    // Note: Resolving promise if the Action was created using a async rawTx
    const rawAction = await this.#rawAction;

    this.#rawAction = await this.#fillRawAction(rawAction);

    const signedTx = await this.#signer.sign(this.#rawAction);

    this.#tx = await this.#provider.sendTransaction(signedTx);
  } catch (error) {
    this._emitErrorEvent(new Error(`Action with error: ${error.message}`));
  }
};
```
Even that way, do you think that worth extract the sign line to a function?

Reply:
Thanksk for explaining!

`fulfill` sounds too much like "to accomplish".

I think `fill` is pretty good given what you're using it for.

Other options that quickly come to mind include `decorateRawAction` or `populateOptionalRawActionFields`

Reply:
Thanks :+1:

Reply:
Can we close this?

Reply:
Sure. Resolving this.